core/src/main/kotlin/io/rover/campaigns/core/events/UserInfo.kt

```diff
@@ -1,44 +1,201 @@
+                    log.w("Deprecation notice: Attempted to remove tags directly instead of using removeTag")
+                    removedTags.forEach { removeTag(it.value) }
+                }
+            } else {
+                log.w("Attempted updating tags to a non collection type")
+                mutableDraft.remove("tags")
+            }
+        }
         currentUserInfo = mutableDraft
     }
 
+    override fun addTag(tag: String) {
+        addTag(tag, null)
+    }
+
+    override fun addTag(tag: String, expiresInSeconds: Long?) {
+        val expires = if (expiresInSeconds != null) {
+            Date(Date().time + expiresInSeconds)
+        } else null
+
+        val mutableTags = currentTags.toMutableSet()
+        mutableTags.add(
+            Tag(
+                value = tag,
+                expires = expires
+            )
+        )
+
+        currentTags = mutableTags
+    }
+
+    override fun removeTag(tag: String) {
+        val mutableTags = currentTags.toMutableSet()
+        mutableTags.remove(Tag(value = tag, expires = null))
+        currentTags = mutableTags
+    }
+
+    override fun tags(): Set<String> {
+        return currentTags.map { it.value }.toSet()
+    }
+
     override fun clear() {
         currentUserInfo = hashMapOf()
+        currentTags = setOf()
+    }
+
+    private var currentTags: Set<Tag> = try {
+        when (store[TAGS_KEY]) {
+            null -> setOf()
+            else -> decodeTags(store[TAGS_KEY]!!)
+        }
+    } catch (throwable: Throwable) {
+        log.w("Corrupted local tags, ignoring and starting fresh. Cause ${throwable.message}")
+        setOf()
     }
+        get() {
+            val now = Date()
+            val unExpiredTags = field.filter {
+                when (it.expires) {
+                    null -> true
+                    else -> it.expires.after(now)
+                }
+            }.toSet()
+
+            if (unExpiredTags != field) {
+                // Some tags have expired update the value
+                currentTags = unExpiredTags
+            }
+            return unExpiredTags
+        }
+        set(value) {
+            field = value
+            store[TAGS_KEY] = encodeTags(value)
+        }
 
     override var currentUserInfo: Attributes = try {
-        val currentAttributesJson = store[USER_INFO_KEY]
-        when (currentAttributesJson) {
+        when (store[USER_INFO_KEY]) {
             null -> hashMapOf()
-            else -> JSONObject(store[USER_INFO_KEY]).toAttributesHash()
+            else -> {
+                val info = JSONObject(store[USER_INFO_KEY]).toAttributesHash().toMutableMap()
+                if (info.containsKey("tags")) {
+                    log.v("Migrating user info tags")
+                    info["tags"]?.let { migrateTags(it) }
+                    info.remove("tags")
+                    store[USER_INFO_KEY] = info.encodeJson().toString()
+                    log.v("Migrated user info tags successfully")
```

Comment:
Did you want to leave these logs here?

Reply:
idk should i? lol 😂. I can remove it.

Reply:
Yeah, i'd remove it, I don't think it adds a ton of value